File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionGroup.java
package br.com.insanitech.spritekit;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by anderson on 06/01/17.
 */

class SKActionGroup extends SKAction {
    private List<SKAction> group = null;

    public SKActionGroup(List<SKAction> group) {
        this.group = group;
    }

    public SKActionGroup(SKActionGroup other) {
        group = other.group;
    }

    @Override
    void computeStart() {
        for (int i = 0; i < group.size(); i++) {
            final SKAction action = group.get(i);
            action.setParent(getParent());
            action.completion = new Runnable() {
                @Override
                public void run() {
                    completedAction(action);
                }
            };
            action.start();
        }
    }

    @Override
    void computeAction(long elapsed) {
        List<SKAction> group = new ArrayList<SKAction>(this.group);
        for (int i = 0; i < group.size(); i++) {
            SKAction action = group.get(i);
            action.computeAction();
        }
    }

    @Override
    void computeFinish() {

    }

    @Override
    boolean willHandleFinish() {
        return true;
    }

    private void completedAction(SKAction action) {
        group.remove(action);
        if (group.size() == 0) {
            getParent().actionCompleted(this);
            dispatchCompletion();
        }
    }
}

File: SpriteKitLib/src/main/java/br/com/insanitech/spritekit/SKActionGroup.kt
package br.com.insanitech.spritekit

import java.util.ArrayList

/**
 * Created by anderson on 06/01/17.
 */

internal class SKActionGroup : SKAction {
    private var group: MutableList<SKAction>? = null

    constructor(group: MutableList<SKAction>) {
        this.group = group
        this.duration = Long.MAX_VALUE
    }

    constructor(other: SKActionGroup) : this(other.group ?: ArrayList<SKAction>())

    internal override fun computeStart() {
        for (i in this.group!!.indices) {
            val action = group!![i]
            action.parent = parent
            action.completion = {
                completedAction(action)
            }
        }
    }

    internal override fun computeAction(elapsed: Long) {
        this.group?.forEach { it.computeAction() }
    }

    internal override fun computeFinish() {

    }

    private fun completedAction(action: SKAction) {
        this.group?.remove(action)
        if (this.group?.size == 0) {
            this.removeFromParent()
            this.dispatchCompletion()
        }
    }
}
